src/libexpr/parser.y

```diff
@@ -686,17 +686,25 @@ Expr * EvalState::parse(
 }
 
 
-SourcePath resolveExprPath(const SourcePath & path)
+SourcePath resolveExprPath(SourcePath path)
 {
+    unsigned int followCount = 0, maxFollow = 1024;
+
     /* If `path' is a symlink, follow it.  This is so that relative
        path references work. */
-    auto path2 = path.resolveSymlinks();
+    while (true) {
+        // Basic cycle/depth limit to avoid infinite loops.
+        if (++followCount >= maxFollow)
+            throw Error("too many symbolic links encountered while traversing the path '%s'", path);
```

Comment:
We may want to say what limit it ran into?

```suggestion
            throw Error("too many symbolic links encountered while traversing the path '%s' (limit is %d)", path, maxFollow);
```

Reply:
Nah, if you have more than 1024 symlinks, it's almost certainly a cycle.

Reply:
Then let's be explicit about intent, then users will know what to do with the error:

```suggestion
            throw Error("more than %d symbolic links encountered while traversing the path '%s' (there is probably a cycle)", path, maxFollow);
```

Reply:
coreutils:

```shellSession
$ ln -s bar foo
$ ln -s foo bar
$ cat foo
cat: foo: Too many levels of symbolic links
```

Of course coreutils is kinda bad with error messages, but we can align with it for a consistent user experience and improve on it.

```suggestion
            throw Error("too many levels of symbolic links while traversing the path '%s'; assuming it's part of a cycle after %d indirections", originalPath, maxFollows);
```

It would also be slightly more helpful to show the original path instead of an arbitrary path within the cycle, in case it's not the cycle that's the mistake, but rather the use of the cycle. Also note that the original symlink does not even have to be part of the cycle; it only has to lead to it.